cli/src/greet_invitation.rs

```diff
@@ -0,0 +1,232 @@
+        greet_invitation.device,
+        |cmds, device| async move {
+            let invitation_token = InvitationToken::from_hex(&greet_invitation.token)
+                .map_err(|e| anyhow::anyhow!(e))?;
+
+            let invitation = step0(&cmds, invitation_token).await?;
+
+            match invitation {
+                InviteListItem::User { .. } => {
+                    let ctx = UserGreetInitialCtx::new(
+                        device,
+                        Arc::new(cmds),
+                        EventBus::default(),
+                        invitation_token,
+                    );
+
+                    let ctx = step1_user(ctx).await?;
+                    let ctx = step2_user(ctx).await?;
+                    let ctx = step3_user(ctx).await?;
+                    let ctx = step4_user(ctx).await?;
+                    step5_user(ctx).await
+                }
+                InviteListItem::Device { .. } => {
+                    let ctx = DeviceGreetInitialCtx::new(
+                        device,
+                        Arc::new(cmds),
+                        EventBus::default(),
+                        invitation_token,
+                    );
+
+                    let ctx = step1_device(ctx).await?;
+                    let ctx = step2_device(ctx).await?;
+                    let ctx = step3_device(ctx).await?;
+                    let ctx = step4_device(ctx).await?;
+                    step5_device(ctx).await
+                }
+            }
+        },
+    )
+    .await
+}
+
+/// Step 0: retrieve info
+async fn step0(
+    cmds: &AuthenticatedCmds,
+    invitation_token: InvitationToken,
+) -> anyhow::Result<InviteListItem> {
+    let handle = start_spinner("Retrieving invitation info");
+
+    let rep = cmds.send(invite_list::Req).await?;
+
+    let invitations = match rep {
+        invite_list::InviteListRep::Ok { invitations } => invitations,
+        rep => return Err(anyhow::anyhow!("Server error: {rep:?}")),
+    };
+
+    let invitation = match invitations.into_iter().find(|invitation| match invitation {
+        InviteListItem::User { token, .. } if *token == invitation_token => true,
+        InviteListItem::Device { token, .. } if *token == invitation_token => true,
+        _ => false,
+    }) {
+        Some(invitation) => invitation,
+        None => return Err(anyhow::anyhow!("Invitation not found")),
+    };
+
+    handle.done();
+
+    Ok(invitation)
+}
+
+/// Step 1: wait peer
+async fn step1_user(ctx: UserGreetInitialCtx) -> anyhow::Result<UserGreetInProgress1Ctx> {
+    let handle = start_spinner("Waiting for claimer");
+
+    let ctx = ctx.do_wait_peer().await?;
+
+    handle.done();
+
+    Ok(ctx)
+}
+
+/// Step 1: wait peer
+async fn step1_device(ctx: DeviceGreetInitialCtx) -> anyhow::Result<DeviceGreetInProgress1Ctx> {
+    let handle = start_spinner("Waiting for claimer");
+
+    let ctx = ctx.do_wait_peer().await?;
+
+    handle.done();
+
+    Ok(ctx)
+}
```

Comment:
All the `stepN_user` functions are almost identical to the `stepN_device`. Isn't there a way to refactor into a single code each time?

I mean, even if they differ in the parameter type (e.g. `UserGreetInitialCtx` vs `DeviceGreetInitialCtx`), these types seem to share the exact same interface.

Reply:
We can use `trait` or `macro` to factorize.
`macro` seems overkill for only `2` calls, because it is less readable
and @touilleMan didn't want to use `trait`